### ISSUE TITLE
Fonts/text are rendered inconsistently in Santa Cruz County COVID-19 shelter-in-place PDF

### ISSUE THREAD

**Issue by dholbert** [mozilla/pdf.js#11840]
Attach (recommended) or Link to PDF file here:
* PDF in original location (link): https://www.santacruzhealth.org/Portals/7/Pdfs/Coronavirus/PHO%20Order%20Extending%20SIP%2020200331.pdf
* PDF as a local github-hosted attachment:
[PHO Order Extending SIP 20200331.pdf](https://github.com/mozilla/pdf.js/files/4518865/PHO.Order.Extending.SIP.20200331.pdf)

Configuration:
- Web browser and its version: Firefox Nightly 77.0a1 (2020-04-21) (64-bit)
- Operating system and its version: Ubuntu 19.10
- PDF.js version: unsure; whatever's currently bundled with Firefox Nightly
- Is a browser extension: I'm using the PDF.js that's built-in to Firefox Nightly (not sure if that counts as an extension; I didn't manually install a different/external version of PDF.js, if that's what this question is asking)

Steps to reproduce the problem:
1. View the attached/linked PDF
2. Look at the font & characters in the page's content (below the header) -- e.g. the section "ORDER OF THE HEALTH OFFICER OF THE COUNTY OF SANTA CRUZ DIRECTING [...]"
3. Compare to a standalone PDF viewer.

What is the expected behavior? (add screenshot)
 - The text (starting with "ORDER OF THE HEALTH OFFICER") **should mostly be displayed in a serif font**.  At least, that's how it's shown in my external viewer, evince. Here's a screenshot of the document up through section "1", from Evince -- note that only some small bits of the header are sans-serif:
![image](https://user-images.githubusercontent.com/426803/80031704-ed59b580-849e-11ea-921e-97f91565ce9f.png)
 

What went wrong? (add screenshot)
- The text **is mostly shown in a sans-serif font**.
- The font seems to change haphazardly. E.g. in the "UNDER THE AUTHORITY...  AND 120175, THE HEALTH OFFICER" paragraph, just before the numbered sections start, the characters up to "AN" are from a sans-serif font, whereas the letter "D" and the following letters are from a serif font. So the font switches midway through the word "AND", which looks quite odd. Screenshot:
![image](https://user-images.githubusercontent.com/426803/80031573-c4d1bb80-849e-11ea-820d-2946e5980d19.png)
- Further down, in `1. This Order`, the "T" and "O" look like they might be from a larger (or wider?) font-size than the other characters in those words. (This seems to be an issue throughout the document, with some letters [mostly capital letters] being slightly wider than expected & feeling a bit unsettling.)  Screenshot: 
![image](https://user-images.githubusercontent.com/426803/80031967-52ada680-849f-11ea-9fa5-9676657ef51c.png)

**Comment by timvandermeij** [comment]
I can't reproduce this using Arch Linux, Firefox 75 and the latest viewer at https://mozilla.github.io/pdf.js/web/viewer.html, however this is most likely because I have installed more fonts. The PDF file does not embed all its fonts, so this is most likely caused by a fallback font being used.

**Comment by dholbert** [comment]
Thanks! For an apples-to-apples comparison: I just checked, and I **am** able to reproduce this on my system, using the latest viewer at https://mozilla.github.io/pdf.js/web/viewer.html , in both Firefox 75 and Firefox Nightly.  So: I suspect you're right about this being a difference in system fonts.

Having said that: do you know why things would still look better in my native PDF viewer, `evince`, though? (I'm assuming it has the same system fonts available as Firefox does, though maybe it has its own special store of backup/PDF-depending fonts...)

Also: as one additional bit of data, Firefox's devtools says that the rendered PDF only uses 3 fonts, for me (though I'm not sure I trust this is accurate[1]): DejaVu Sans, DejaVu Sans Mono, and Ubuntu:
![image](https://user-images.githubusercontent.com/426803/80039043-bdfd7580-84ab-11ea-955e-08d14faa0f93.png)

[1] (I'm not sure I trust our devtools-reported fonts here because both "AN" and "D" are reported as using DejaVu Sans, when clearly "D" is actually drawn with serifs, as shown in my initial screenshot. It looks like the reported fonts are for PDF.js's transparent "text layer" which is separate from the actual painted characters in the "canvas layer"; so I don't know if these text layer fonts are relevant here or not.

**Comment by timvandermeij** [comment]
I'm not too familiar with the font conversion code, but the PDF specification describes 14 standard fonts that PDF viewers may use for font substitution. I think that's what Evince (or Okular in my case) use. Refer to #11637 for more information. Basically PDF.js requires fonts to be embedded, or it will use a fallback font.

**Comment by dholbert** [comment]
Makes sense. Thanks!

**Comment by marco-c** [comment]
#16363 might have fixed this, can you still reproduce in latest Firefox Nightly?

**Comment by dholbert** [comment]
Can't reproduce in latest Firefox Nightly, no.  Here's a screenshot of what I see now for the bad sections shown at the end of my initial comment here:
![image](https://github.com/mozilla/pdf.js/assets/426803/b08599a3-5747-4e9c-b390-14ec830bf0dc)

The font looks to be consistently the same serif font now.

**Comment by dholbert** [comment]
Hmm, I actually also can't reproduce when using Firefox Nightly 2020-04-21 (the version that I was testing when I reported this bug), launched via mozregression.

I'm on Ubuntu 22.04 now, on a different machine from the machine that I used to file this bug.  (I don't have the original system anymore.)  Not sure offhand what the variable might be that's preventing me from reproducing; maybe I've got different fonts available now, or there was an Ubuntu bug-fix that influenced how we choose fonts, or something else.

In any case, since I can't reproduce, I can't be entirely sure whether the bug is still present or not.  I'm happy to have this closed as no-longer-reproducible (and maybe even fixed) if that makes sense.
